accounting_service/models.py

```diff
@@ -299,3 +374,218 @@ def __repr__(self):
+        # when we do comparisons.
+        #
+        # This harmonizes this.
+        return self.sample_time.astimezone(timezone.utc)
+
+    __table_args__ = (
+        Index(
+            "billableresourceconsumptionratesample_workspace_time_index",
+            "workspace",
+            "sample_time",
+        ),
+    )
+
+    @classmethod
+    def insert_from_message(
+        cls, session: Session, msg: eodhp_utils.pulsar.messages.BillingResourceConsumptionRateSample
+    ) -> Optional[UUID]:
+        result = session.execute(
+            insert(cls)
+            .values(
+                uuid=UUID(msg.uuid),
+                sample_time=datetime_default_to_utc(datetime.fromisoformat(msg.sample_time)),
+                item_id=(
+                    select(BillingItem.uuid).where(BillingItem.sku == msg.sku).scalar_subquery()
+                ),
+                user=UUID(msg.user) if msg.user else None,
+                workspace=msg.workspace,
+                rate=msg.rate,
+            )
+            .on_conflict_do_nothing(index_elements=["uuid"])
+            .returning(cls.uuid)
+        )
+
+        return result.scalar_one_or_none()
+
+    @classmethod
+    def find_data_for_interval(
+        cls, session: Session, workspace: str, sku: str, start: datetime, end: datetime
+    ) -> Sequence[Self]:
+        item_subquery = select(BillingItem.uuid).where(BillingItem.sku == sku).scalar_subquery()
+        last_before_start = (
+            select(cls)
+            .where(cls.item_id == item_subquery)
+            .where(cls.workspace == workspace)
+            .where(cls.sample_time <= start)
+            .order_by(cls.sample_time.desc())
+            .limit(1)
+        )
+        first_after_end = (
+            select(cls)
+            .where(cls.item_id == item_subquery)
+            .where(cls.workspace == workspace)
+            .where(cls.sample_time >= end)
+            .order_by(cls.sample_time)
+            .limit(1)
+        )
+        in_period = (
+            select(cls)
+            .where(cls.item_id == item_subquery)
+            .where(cls.workspace == workspace)
+            .where(cls.sample_time > start)
+            .where(cls.sample_time < end)
+        )
+
+        if db.settings.SQL_DRIVER.startswith("sqlite"):
+            # Only used in tests (but the tests can be run with PostgreSQL as well).
+            # SQLite can't cope with the UNION syntax used by SQLAlchemy.
+            return (
+                list(session.execute(last_before_start).scalars().all())
+                + list(session.execute(in_period).scalars().all())
+                + list(session.execute(first_after_end).scalars().all())
+            )
+
+        query = select(cls).from_statement(
+            union(last_before_start, first_after_end, in_period).order_by("sample_time")
+        )
+        return session.execute(query).scalars().all()
+
+    @classmethod
+    def calculate_consumption_for_interval(
```

Comment:
Is there a scenario where we would have a single-sample, would that cause an `IndexError`?